### PR TITLE
Return "None" for unseen values in categorical dicts.

### PR DESCRIPTION
**Pull request by minikomi**
### What

This is a breaking change updating how categorical dicts are used.
Version updates to `0.3.0`.

- You now MUST pass a list of categorical columns expected to be
  covered by the categorical dicts.
- Any "unseen" values will return `None` by default. Setting a `default_unseen` value is also supported.

### Why

- This is more in line with how categorically transformed columns are used in practice
  eg. [LightGBM](https://lightgbm.readthedocs.io/en/latest/Advanced-Topics.html#missing-value-handle)

### CODE REVIEW

File: carling/categorical.py [hunk 1]
```diff
@@ -137,22 +137,32 @@ def CreateCategoricalDicts(pcoll, cat_cols, existing_dict_rows):
...
+    categorical_mapping_dicts = (
+        categorical_dict_rows
+        | "index by column" >> beam.Map(lambda r: (r[0], (r[1], r[2])))
+        | "group by column" >> beam.GroupByKey()
+        | "create dict per column" >> beam.Map(lambda r: (r[0], dict(r[1])))
     )
 
+    def _process(row, d):
+        for col in cat_cols:
+            row[col] = d.get(col, {}).get(row[col], default_unseen)
```
Comment (wafrelka) [inline carling/categorical.py:162]: According to the Apache Beam's document, element should not be mutated.
It is better to create a new dictionary instead.

reference: <https://beam.apache.org/releases/pydoc/2.27.0/apache_beam.transforms.core.html#apache_beam.transforms.core.DoFn.process>

File: tests/test_categorical.py [hunk 2]
```diff
@@ -86,6 +86,58 @@ def test(self):
 ]
 
 
+class ReplaceCategoricalColumnsTest(unittest.TestCase):
```
Comment (wafrelka) [inline tests/test_categorical.py:89]: It is better to make sure that non-categorical columns are not updated in `ReplaceCategoricalColumns`.

File: tests/test_categorical.py [hunk 3]
```diff
@@ -86,6 +86,58 @@ def test(self):
...
+
+        expected = [
+            # Seen values
+            {"key": 1, "column1": 10, "column2": 2, "column3": None},
+            # Unseen values
+            {"key": 2, "column1": None, "column2": None, "column3": None},
+            # Mixed
+            {"key": 3, "column1": None, "column2": 3, "column3": None},
+            # Repeat Row
+            {"key": 4, "column1": None, "column2": 3, "column3": None},
```
Comment (wafrelka) [inline tests/test_categorical.py:100]: What is the intention of testing repeated rows?

File: carling/categorical.py (hunk 1 above)
Comment (minikomi) [reply carling/categorical.py]: 974418e

File: tests/test_categorical.py (hunk 2 above)
Comment (minikomi) [reply tests/test_categorical.py:89]: Testing repeat substitutions give the same result for the same values.

File: tests/test_categorical.py (hunk 3 above)
Comment (minikomi) [reply tests/test_categorical.py:100]: Testing repeat substitutions give the same result for the same values.

File: tests/test_categorical.py (hunk 2 above)
Comment (minikomi) [reply tests/test_categorical.py:89]: fc02ca8 @wafrelka

**Review by wafrelka — APPROVED** [review]
Overall I'm fine with this change, but I feel the testcases are a little complicated.

**Comment by minikomi** [comment]
Agreed. Should split into separate cases in future.

**Pull request merged**